book/src/libraries/api_lib_guide.md

```diff
@@ -0,0 +1,319 @@
+# API guidelines
+
+## 1. Introduction
+
+- **Purpose**: Provide plc library developers with information how an interface for an application written in `IEC61131-3` should be designed and why.
+- **Scope**: This guideline applies to all developers writing libraries for
+         use in IEC61131-3 applications.
+
+## 2. API Guidelines
+
+### 2.1 `VAR_IN_OUT` instead of pointers
+
+If a function takes a parameter for the purpose of reading and writing to it
+a `VAR_IN_OUT` can be used instead of a pointer in the `VAR_INPUT`.
+
+### 2.2 `FUNCTION` and `FUNCTION_BLOCK`
+
+`FUNCTION` and `FUNCTION_BLOCK` have similar properties, but they have fundamentally different representation in the compiler.
+A `FUNCTION` is defined in a similar manner to a `C` function:
+- It has no backing struct
+- values defined inside it will only survive for the duration of the function call
+
+Example:
+
+```iecst
+FUNCTION myFunc : DINT
+VAR_INPUT
+  x  : DINT;
+END_VAR
+END_FUNCTION
+```
+```c
+int32_t myFunc(int32_t);
+```
+
+In contrast, a `FUNCTION_BLOCK` is backed by a struct and is globally accessible by a defined instance.
+To declare a `FUNCTION_BLOCK`, a backing struct has to be declared and passed as a reference to the function block implementation.
+
+```iecst
+FUNCTION_BLOCK myFb
+VAR_INPUT
+  x  : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+typedef struct {
+  int32_t x;
+} myFunctStr;
+
+void myFb(myFunctStr*);
+```
+
+
+
+#### 2.2.1 Parameters
+
+`FUNCTION` and `FUNCTION_BLOCK` may define input parameters. These are passed using the `VAR_INPUT` or `VAR_IN_OUT` blocks.
+The difference between the two blocks is how the values are passed.
+A `VAR_INPUT` variable is passed _by value_, while a `VAR_IN_OUT` variable is passed _by reference_.
+In general, it is recommended to use a `VAR_IN_OUT` for data that needs to be both read and written, while `VAR_INPUT` should be reserved to read only values.
+
+> **_NOTE:_** In `FUNCTION`s complex datatypes are handled as pointers.
```

Comment:
is this the case for things other than VLAs?

Reply:
@ghaith can you please help here?

Reply:
I guess it is correct as in they are passed as pointers, but they are then deep-copied and the passed arguments are therefore immutable. I think 'handled as pointers' could be misleading here, but I'm struggling to come up with a concise alternative.

Reply:
We should probably describe more of what it means to be handled as pointers. For the target audience here, (A C/C++ Developer) the function signature should be a pointer, but they should still know that the value being passed is actually copied.
I think we mention this elsewere as well. Not sure how to formulate it best.

Reply:
regarding VLA @mhasel  it's for all complex times now, we did this for performance.